### PR TITLE
Fix topo complex

### PR DESCRIPTION
**Pull request by dlyr**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Bug fix

* **What is the current behavior?** (You can also link to an open issue here)
When converting to Topological Meshes, some non manifold faces are detected, some raise an assert (on purpose according to openmesh doc, see commit messages).

* **What is the new behavior (if this is a feature change)?**
We pre process the index list to skip badly formed faces.
For the current version, it removes the faces (since inputs are only triangles, bad formed faces drops down to lines which are not handled)
When #589 will be merged, it can add faces that "loose" some vertices, but are still relevant.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
nope

* **Other information**:

### CODE REVIEW

**Review by nmellado — COMMENTED** [review]
I don't have time for an in-depth review, but my first feeling is that your code is not documented enough: I have seen you refer to another documentation, but in current state it is a bit difficult to follow the code

File: src/Core/Geometry/TopologicalMesh.inl [hunk 1]
```diff
@@ -111,9 +111,72 @@ inline TopologicalMesh::TopologicalMesh( const TriangleMesh& triMesh,
...
+                    end--;
+                    // explicit del wedge (not needed in initWithWedges)
+                    m_wedges.del( *wedgeEnd );
+                    wedgeEnd--;
+                    normalEnd--;
+                }
+                face_vhandles.erase( end + 1, face_vhandles.end() );
+                face_wedges.erase( wedgeEnd + 1, face_wedges.end() );
+                face_normals.erase( normalEnd + 1, face_normals.end() );
+            }
```
[the remark this answers is not in the document]
Comment (dlyr) [reply src/Core/Geometry/TopologicalMesh.inl:135]: Well, it's line 293 just below
The idea is to remove consecutive references to the same topological vertex
for instance if your face vhandles is 1 2 2, it is not valid on the OpenMesh side.
So I remove duplication in vhandles and update other vectors (normal and wedges) accordingly.
First I remove front / back equal vhandles, for instance in the case 1 2 1
Then I remove consecutive for instance 1 2 2 
This code is already adapted to poly mesh
so 1 1 1 2 2 3 3 3 1 1 becomes 1 2 3

**Comment by dlyr** [comment]
The proposed algo will also work on TopologicalMesh for PolyMesh, in this case the first vertex wedge is used.
wedge-degen

**Review by nmellado — APPROVED** [review]

**Comment by nmellado** [comment]
Good for me, once updated wrt `master`

**Comment by dlyr** [comment]
Let me write some documentation and rebase.

File: doc/developer/mesh.md [hunk 2]
```diff
@@ -109,6 +109,14 @@ During manipulation, if two wedges become the same (i.e. for the example on the
 To merge wedges an explicit call to TopologicalMesh::mergeEqualWedges is needed.
 When a wedge is not referenced anymore (because referencing halfedges have been deleted), it is marked for deletation. When one call Ra::Core::Geometry::TopologicalMesh::garbage_collection(), the marked wedges are eventually deleted and halfedges index are updated accordingly.
 
+During convertion from Core::Geometry::*Mesh to Core::Geometry::TopologicalMesh, non manifold mesh faces are not added but given to a optionally provided user defined functor (\see Ra::Core::Geometry::TopologicalMesh::TopologicalMesh).
```
[the remark this answers is not in the document]
Comment (MathiasPaulin) [reply doc/developer/mesh.md:112]: Even if documented in the source, saying here what is the main purpose of such a functor might be helpful.

File: doc/developer/mesh.md [hunk 3]
```diff
@@ -109,6 +109,14 @@ During manipulation, if two wedges become the same (i.e. for the example on the
...
+Some meshes have multiple times the **same** vertices with **different** values for attributes inside a single face.
```
Comment (MathiasPaulin) [inline doc/developer/mesh.md:113]: And ? What is the bad effect of such a case? Given a cube, vertices have the same position but with 3 attributes (one par square face). And this situation is not problematic as wedges are built for that, right ?

File: doc/developer/mesh.md [hunk 4]
```diff
@@ -109,6 +109,14 @@ During manipulation, if two wedges become the same (i.e. for the example on the
...
+This kind of situation can either raise degerated faces, (in case of triangle mesh) or valid faces for some cases for polymesh (\todo will be supported when TopologicalMesh will support PolyMesh).
```
[the remark this answers is not in the document]
Comment (MathiasPaulin) [reply doc/developer/mesh.md:114]: is it only "one vertex, several attributes" described above that gives this "kind of situation" ?

File: doc/developer/mesh.md [hunk 5]
```diff
@@ -109,6 +109,14 @@ During manipulation, if two wedges become the same (i.e. for the example on the
...
+In the latter case, the face is cleaned from the duplicated position and added to the TopologicalMesh, using the first encountered wedge value.
```
Comment (MathiasPaulin) [inline doc/developer/mesh.md:115]: Why the first. Is there a way, for polymesh, to specify which wedge to take (e.g. The one with the maximum cotangent weight) ?

File: doc/developer/mesh.md (hunk 3 above)
Comment (dlyr) [reply doc/developer/mesh.md:113]: inside a single face

File: doc/developer/mesh.md (hunk 5 above)
Comment (dlyr) [reply doc/developer/mesh.md:115]: arbitrarly ... their is no cotan weight here, the position are the same .... 
the other solution is to drop them to the functor for non manifold faces.

File: doc/developer/mesh.md (hunk 3 above)
Comment (dlyr) [reply doc/developer/mesh.md:113]:
```suggestion
Some meshes have multiple times the **same** vertices with **different** values for attributes inside a **single** face.
```

File: doc/developer/mesh.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlyr) [reply doc/developer/mesh.md]: Their is no purpose, it's user defined.

File: doc/developer/mesh.md (hunk 3 above)
Comment (dlyr) [reply doc/developer/mesh.md]: Also there is a figure to show the problem, it's really related to bad topology, for instance if the top face of the cube is a pentagon with two time the same position with different normals. The rendering will change dramatically depending on triangulation, and it's bad ;)

File: doc/developer/mesh.md (hunk 5 above)
Comment (dlyr) [reply doc/developer/mesh.md:118]: see figure (here in the comment, or [updated in the doc](https://github.com/STORM-IRIT/Radium-Engine/blob/f7e57ca4e7c5ca01cc33e61433a12678f1aa72b6/doc/images/wedge-degen.svg))

**Review by MathiasPaulin — APPROVED** [review]

**Pull request merged**